zookeeper-server/src/main/java/org/apache/zookeeper/server/PrepRequestProcessor.java

```diff
@@ -1003,6 +1004,32 @@ public static List<ACL> fixupACL(String path, List<Id> authInfo, List<ACL> acls)
             throw new KeeperException.InvalidACLException(path);
         }
         List<ACL> rv = new ArrayList<>();
+
+        // Overwrite the acl list for users (except super user) when znode group acl feature is on
+        // Set znode ACL to the corresponding znode group acl designated ACL instead
+        // For single domain user / cross domain components -> (x509 : domainName)
+        // For user w/o a domain -> (x509: clientURI)
+        if (X509AuthenticationConfig.getInstance().isX509ClientIdAsAclEnabled() && ProviderRegistry
+            .getServerProvider(
+                X509AuthenticationUtil.X509_SCHEME) instanceof X509ZNodeGroupAclProvider
+            && !X509AuthenticationConfig.getInstance().isDedicatedServerEnabled()) {
+            authInfo.forEach(id -> {
+                if (id.getScheme().equals(X509AuthenticationUtil.X509_SCHEME)) {
```

Comment:
Looks like those if's could be combined.

Reply:
@mgao0 will make a call here.

Reply:
See above